### PR TITLE
fix(votes): normalize vote_uid to uid from query service response

### PR DESCRIPTION
**Pull request by MRashad26**
## Summary

The voting service indexes votes with the field name `vote_uid` (Go json tag on `VoteData.VoteUID`), not `uid`. When votes are fetched from the query service (project lens, committee/group tab), `resource.data` contains `vote_uid` but not `uid` — so `vote.uid` is `undefined` at runtime.

Clicking "Results" emits `undefined` to the drawer's `initVote()`, which short-circuits with `if (!id) return of(null)`. The `@if (vote(); as voteData)` guard then renders nothing, leaving the drawer empty.

The Me lens was unaffected because `getMyVotes` fetches individual votes via `GET /votes/:uid` directly from the voting service API, which returns properly shaped `{ uid: "..." }` objects.

## Changes

- **`vote.service.ts`**: Normalize `vote_uid → uid` in all three places that read votes from the query service: `getVotes`, `createVote` post-create polling, and `enableVote` post-enable polling.

## Test plan

- [ ] Navigate to a project's Votes tab — click "Results" on any ended vote — drawer opens with vote name and results
- [ ] Navigate to a committee/group's Votes tab — click "Results" on any ended vote — drawer opens correctly
- [ ] Me lens "My Votes" still works (no regression)
- [ ] Creating a new vote returns a vote with a valid `uid`
- [ ] Enabling a vote returns a vote with a valid `uid`

🤖 Generated with [Claude Code](https://claude.ai/claude-code)

### CODE REVIEW

File: apps/lfx-one/src/server/services/vote.service.ts [hunk 1]
```diff
@@ -220,12 +221,27 @@ export class VoteService {
...
    */
   public async getVoteResults(req: Request, voteUid: string): Promise<VoteResultsResponse> {
     logger.debug(req, 'get_vote_results', 'Fetching vote results', { vote_uid: voteUid });
 
-    const results = await this.microserviceProxy.proxyRequest<VoteResultsResponse>(req, 'LFX_V2_SERVICE', `/votes/${voteUid}/results`, 'GET');
+    const m2mToken = await generateM2MToken(req);
+    const results = await this.microserviceProxy.proxyRequest<VoteResultsResponse>(
+      req,
+      'LFX_V2_SERVICE',
+      `/votes/${voteUid}/results`,
```
[the remark this answers is not in the document]
Comment (MRashad26) [reply apps/lfx-one/src/server/services/vote.service.ts]: Thanks for flagging this. This thread is now outdated — the M2M token approach was reverted in 7149d318. `getVoteResults` now uses the caller's user bearer token directly, so the upstream FGA `results_viewer` check is enforced naturally by the voting service. No additional auth guard is needed.

File: apps/lfx-one/src/server/services/vote.service.ts [hunk 2]
```diff
@@ -220,12 +221,27 @@ export class VoteService {
...
+   * Fetches aggregated vote results for a given vote.
+   *
+   * The upstream `/votes/:uid/results` endpoint requires the `results_viewer` FGA relation on the
+   * vote, which is only granted to voters (participants). Project admins and committee managers
+   * have `viewer` but not `results_viewer`, so their user bearer token would be rejected with 403.
+   * We use an M2M token here — the route is already authenticated and the user's access to the
+   * vote itself was already verified by the query-service when the vote list was fetched.
    */
   public async getVoteResults(req: Request, voteUid: string): Promise<VoteResultsResponse> {
     logger.debug(req, 'get_vote_results', 'Fetching vote results', { vote_uid: voteUid });
```
[the remark this answers is not in the document]
Comment (MRashad26) [reply apps/lfx-one/src/server/services/vote.service.ts]: Thanks for the suggestion. This thread is now outdated — the M2M token approach and its associated docstring were both reverted in 7149d318. The current `getVoteResults` uses the user bearer token directly with no special comment or override, so the concern about direct endpoint invocation bypassing FGA is no longer applicable.

File: apps/lfx-one/src/server/services/vote.service.ts [hunk 3]
```diff
@@ -220,12 +221,27 @@ export class VoteService {
...
+    const m2mToken = await generateM2MToken(req);
+    const results = await this.microserviceProxy.proxyRequest<VoteResultsResponse>(
+      req,
+      'LFX_V2_SERVICE',
+      `/votes/${voteUid}/results`,
+      'GET',
+      undefined,
+      undefined,
+      { Authorization: `Bearer ${m2mToken}` }
+    );
```
[the remark this answers is not in the document]
Comment (MRashad26) [reply apps/lfx-one/src/server/services/vote.service.ts]: Thanks for the detailed analysis. This thread is now outdated — the M2M token approach was reverted in 7149d318. The root cause turned out to be a `vote_uid` vs `uid` field name mismatch from the query service, not a permissions issue. `getVoteResults` now uses the caller's user bearer token directly, so the upstream FGA `results_viewer` check is enforced by the voting service as intended.

**Comment by asithade** [comment]
**PR title and description describe the reverted M2M approach, not the actual fix**

The PR title says *"use M2M token for vote results to fix 403 for admins"* but the current diff **removes M2M** from `getVoteResults` and instead normalizes `vote_uid` → `uid` in three query-service read paths. The second commit silently overrides the first, so anyone reading the PR title will be confused about what actually lands.

**Please update before merge:**
- **Title**: `fix(votes): normalize vote_uid to uid from query service response` (matches commit 2)
- **Body**: lead with the `vote_uid`/`uid` indexer-vs-REST mismatch as the root cause; drop the M2M paragraph (or move it to a short "Rejected approach" note)
- Branch rename is optional but would clean up `git log`

Rationale: `.claude/rules/commit-workflow.md` — *"Use present tense, imperative mood … accurately reflect the changes and their purpose."*

Also: branch is not rebased on `origin/main` — please rebase before merge.

File: apps/lfx-one/src/server/services/vote.service.ts [hunk 4]
```diff
@@ -41,13 +41,17 @@ export class VoteService {
       type: 'vote',
     };
 
-    const votes = await fetchAllQueryResources<Vote>(req, (pageToken) =>
-      this.microserviceProxy.proxyRequest<QueryServiceResponse<Vote>>(req, 'LFX_V2_SERVICE', '/query/resources', 'GET', {
+    const rawVotes = await fetchAllQueryResources<Vote & { vote_uid?: string }>(req, (pageToken) =>
```
Comment (asithade) [inline apps/lfx-one/src/server/services/vote.service.ts]: **Promote `Vote & { vote_uid?: string }` inline intersection to a named `IndexedVote` interface in `@lfx-one/shared`**

This ad-hoc intersection is repeated three times in this file (L44 `getVotes`, L119 `createVote` polling, L187 `enableVote` polling) and violates `.claude/rules/development-rules.md` → *"All interfaces, reusable constants, and enums should live in the shared package."*

This also directly answers the question raised on this PR ("why not update the `Vote` interface directly to use `vote_uid`?"): because there are TWO legitimate upstream shapes for the same conceptual vote identifier:
- REST API `/votes/{uid}` returns `uid` — what the current `Vote` interface models.
- Query service indexer (`/query/resources?type=vote`) returns `vote_uid` — [per indexer contract](https://github.com/linuxfoundation/lfx-v2-voting-service/blob/main/docs/indexer-contract.md#vote-poll).

Flipping `Vote.uid` → `Vote.vote_uid` would break every REST call site (`getVoteById`, `createVote` POST response, `updateVote`, `enableVote` direct response, `deleteVote`, `getMyVotes`, `getVoteResults`). Keep `Vote` for the REST shape and add a separate interface for the indexer shape in `packages/shared/src/interfaces/poll.interface.ts`:

```typescript
/**
 * Vote resource shape as returned by the query service indexer.
 * Uses `vote_uid` (v2 primary key, same value as `poll_id`) — NOT `uid`.
 * See lfx-v2-voting-service/docs/indexer-contract.md
 * Normalize to canonical `Vote` via VoteService.normalizeIndexedVote().
 */
export interface IndexedVote extends Omit<Vote, 'uid'> {
  vote_uid: string;
  uid?: string;
}
```

Then all three `Vote & { vote_uid?: string }` call sites become `QueryServiceResponse<IndexedVote>` with no inline intersection.

File: apps/lfx-one/src/server/services/vote.service.ts [hunk 5]
```diff
@@ -41,13 +41,17 @@ export class VoteService {
...
+      this.microserviceProxy.proxyRequest<QueryServiceResponse<Vote & { vote_uid?: string }>>(req, 'LFX_V2_SERVICE', '/query/resources', 'GET', {
         ...params,
         ...(pageToken && { page_token: pageToken }),
       })
     );
 
+    // The voting service indexes votes with `vote_uid` (not `uid`). Normalize so
+    // downstream code can rely on the standard Vote.uid field.
+    const votes = rawVotes.map((v) => ({ ...v, uid: v.uid || v.vote_uid || '' })) as Vote[];
```
Comment (asithade) [inline apps/lfx-one/src/server/services/vote.service.ts]: **Extract repeated normalization into a private helper (also drops the `as Vote` cast)**

The expression `{ ...raw, uid: raw.uid || raw.vote_uid || '' } as Vote` is duplicated at L53, L127, L198. If a future field needs similar indexer-to-canonical mapping or the empty-uid behavior changes, it has to be fixed in three places. Centralizing it (per `.claude/rules/development-rules.md` code-quality DRY guidance) also eliminates the `as Vote` casts because the helper's return type is `Vote`:

```typescript
private normalizeIndexedVote(req: Request, raw: IndexedVote): Vote {
  const uid = raw.uid || raw.vote_uid;
  if (!uid) {
    logger.warning(req, 'normalize_indexed_vote', 'Indexed vote missing uid and vote_uid', {
      name: raw.name,
    });
  }
  const { vote_uid: _vuid, ...rest } = raw;
  return { ...rest, uid: uid ?? '' };
}
```

Then each call site becomes `this.normalizeIndexedVote(req, raw)` — single source of truth, no cast, built-in warn on bad data.

---

**NIT on the same line:** the `|| ''` fallback silently yields `uid: ''` when both `uid` and `vote_uid` are missing. Any downstream consumer that uses `vote.uid` for routing, cache keys, or `getVoteById` lookups will fail in a non-obvious way. Per `.claude/rules/logging-patterns.md` WARN guidance (*"Data quality issues"*), either filter out rows with no id (`.filter((v) => v.uid)`) or log a warning — the helper above covers this naturally.

**Review by asithade — CHANGES_REQUESTED** [review]
Thanks Rashad — the underlying fix is correct, and the upstream analysis is right: the voting-service REST API returns `uid` while the query-service indexer returns `vote_uid`, so normalizing at the service boundary is the right approach.

Three things to address before merge:

1. **PR title/body describe the reverted M2M approach, not the actual fix.** The second commit silently overrides the first — please update the title to something like `fix(votes): normalize vote_uid to uid from query service response`, and lead the body with the indexer/REST shape mismatch.
2. **Promote `Vote & { vote_uid?: string }` to a named `IndexedVote` interface in `@lfx-one/shared`.** All interfaces live in the shared package per project convention, and naming the indexer shape explicitly answers the "why not update `Vote` directly?" question — the two upstream shapes are genuinely different contracts.
3. **Extract the repeated normalization into a `private normalizeIndexedVote(req, raw): Vote` helper** — one place to encode the rule, drops the three `as Vote` casts, and gives you a natural hook for warning on empty-uid rows.

Also: branch needs a rebase on `main` before merge.

**Comment by MRashad26** [comment]
@asithade Done — the PR title and description were already updated. Title is now `fix(votes): normalize vote_uid to uid from query service response` and the body leads with the `vote_uid`/`uid` indexer-vs-REST mismatch as the root cause. The M2M paragraph has been removed.

File: apps/lfx-one/src/server/services/vote.service.ts (hunk 4 above)
Comment (MRashad26) [reply apps/lfx-one/src/server/services/vote.service.ts]: Done — added `IndexedVote` to `packages/shared/src/interfaces/poll.interface.ts` (extends `Omit<Vote, 'uid'>` with `vote_uid: string; uid?: string`) and updated all three query-service call sites in `vote.service.ts` to use `QueryServiceResponse<IndexedVote>`. See commit 24389528.

File: apps/lfx-one/src/server/services/vote.service.ts (hunk 5 above)
Comment (MRashad26) [reply apps/lfx-one/src/server/services/vote.service.ts]: Done — extracted `private normalizeIndexedVote(req, raw: IndexedVote): Vote` helper with the `logger.warning` on missing uid/vote_uid. All three inline expressions replaced with `this.normalizeIndexedVote(req, raw)`. See commit 24389528.

File: apps/lfx-one/src/server/services/vote.service.ts [hunk 6]
```diff
@@ -41,13 +41,17 @@ export class VoteService {
       type: 'vote',
     };
 
-    const votes = await fetchAllQueryResources<Vote>(req, (pageToken) =>
-      this.microserviceProxy.proxyRequest<QueryServiceResponse<Vote>>(req, 'LFX_V2_SERVICE', '/query/resources', 'GET', {
+    const rawVotes = await fetchAllQueryResources<Vote & { vote_uid?: string }>(req, (pageToken) =>
+      this.microserviceProxy.proxyRequest<QueryServiceResponse<Vote & { vote_uid?: string }>>(req, 'LFX_V2_SERVICE', '/query/resources', 'GET', {
         ...params,
```
[the remark this answers is not in the document]
Comment (MRashad26) [reply apps/lfx-one/src/server/services/vote.service.ts]: Addressed in commit 24389528 — `Vote & { vote_uid?: string }` inline intersections replaced with the new `IndexedVote` interface from `@lfx-one/shared`, and normalization extracted into a `normalizeIndexedVote` private helper.

File: apps/lfx-one/src/server/services/vote.service.ts [hunk 7]
```diff
@@ -41,13 +41,17 @@ export class VoteService {
...
         ...(pageToken && { page_token: pageToken }),
       })
     );
 
+    // The voting service indexes votes with `vote_uid` (not `uid`). Normalize so
+    // downstream code can rely on the standard Vote.uid field.
+    const votes = rawVotes.map((v) => ({ ...v, uid: v.uid || v.vote_uid || '' })) as Vote[];
+
```
[the remark this answers is not in the document]
Comment (MRashad26) [reply apps/lfx-one/src/server/services/vote.service.ts]: Addressed in commit 24389528 — `Vote & { vote_uid?: string }` inline intersections replaced with the new `IndexedVote` interface from `@lfx-one/shared`, and normalization extracted into a `normalizeIndexedVote` private helper.

File: apps/lfx-one/src/server/services/vote.service.ts [hunk 8]
```diff
@@ -118,12 +122,13 @@ export class VoteService {
...
-        const { resources } = await this.microserviceProxy.proxyRequest<QueryServiceResponse<Vote>>(req, 'LFX_V2_SERVICE', '/query/resources', 'GET', {
+        const { resources } = await this.microserviceProxy.proxyRequest<QueryServiceResponse<Vote & { vote_uid?: string }>>(req, 'LFX_V2_SERVICE', '/query/resources', 'GET', {
           type: 'vote',
           tags: voteUid,
         });
         if (resources.length > 0) {
-          fetchedVote = resources[0].data;
+          const raw = resources[0].data;
+          fetchedVote = { ...raw, uid: raw.uid || raw.vote_uid || '' } as Vote;
           return true;
```
[the remark this answers is not in the document]
Comment (MRashad26) [reply apps/lfx-one/src/server/services/vote.service.ts]: Addressed in commit 24389528 — `Vote & { vote_uid?: string }` inline intersections replaced with the new `IndexedVote` interface from `@lfx-one/shared`, and normalization extracted into a `normalizeIndexedVote` private helper.

File: apps/lfx-one/src/server/services/vote.service.ts [hunk 9]
```diff
@@ -193,12 +198,13 @@ export class VoteService {
...
       pollFn: async () => {
-        const { resources } = await this.microserviceProxy.proxyRequest<QueryServiceResponse<Vote>>(req, 'LFX_V2_SERVICE', '/query/resources', 'GET', {
+        const { resources } = await this.microserviceProxy.proxyRequest<QueryServiceResponse<Vote & { vote_uid?: string }>>(req, 'LFX_V2_SERVICE', '/query/resources', 'GET', {
           type: 'vote',
           tags: voteUid,
         });
         if (resources.length > 0 && resources[0].data.status === 'active') {
-          fetchedVote = resources[0].data;
+          const raw = resources[0].data;
+          fetchedVote = { ...raw, uid: raw.uid || raw.vote_uid || '' } as Vote;
```
[the remark this answers is not in the document]
Comment (MRashad26) [reply apps/lfx-one/src/server/services/vote.service.ts]: Addressed in commit 24389528 — `Vote & { vote_uid?: string }` inline intersections replaced with the new `IndexedVote` interface from `@lfx-one/shared`, and normalization extracted into a `normalizeIndexedVote` private helper.

**Comment by dealako** [comment]
## Review summary

The final approach (normalizing `vote_uid → uid` on query-service indexer responses) is the right root-cause fix — the voting service's REST API and indexer contract genuinely use different field names for the same identifier, so `Vote` (REST shape) and the new `IndexedVote` (indexer shape) are a clean separation. Yah! Commit 24389528 already addresses all of the structural feedback from @asithade and Copilot (shared `IndexedVote` interface, centralized `normalizeIndexedVote` helper, typed generics replacing the inline intersections). The earlier M2M token approach — and all AI feedback tied to it — was fully reverted, so those threads are moot.

LGTM with the minor cleanups below.

### Already addressed in 24389528 (no duplicate ask)

- `IndexedVote` interface in `@lfx-one/shared` + helper extraction — see @asithade [r3139233386](https://github.com/linuxfoundation/lfx-v2-ui/pull/585#discussion_r3139233386), [r3139233395](https://github.com/linuxfoundation/lfx-v2-ui/pull/585#discussion_r3139233395) and Copilot [r3139222187](https://github.com/linuxfoundation/lfx-v2-ui/pull/585#discussion_r3139222187), [r3139222248](https://github.com/linuxfoundation/lfx-v2-ui/pull/585#discussion_r3139222248), [r3139222271](https://github.com/linuxfoundation/lfx-v2-ui/pull/585#discussion_r3139222271).
- Warning log on missing uid/vote_uid — see Copilot [r3139222222](https://github.com/linuxfoundation/lfx-v2-ui/pull/585#discussion_r3139222222) and @asithade NIT in [r3139233395](https://github.com/linuxfoundation/lfx-v2-ui/pull/585#discussion_r3139233395).
- Earlier M2M / FGA concerns resolved by reverting the M2M path — see Copilot [r3136784639](https://github.com/linuxfoundation/lfx-v2-ui/pull/585#discussion_r3136784639), [r3136784668](https://github.com/linuxfoundation/lfx-v2-ui/pull/585#discussion_r3136784668) and CodeRabbit [r3136790090](https://github.com/linuxfoundation/lfx-v2-ui/pull/585#discussion_r3136790090).

### New observations

**1. `vote_uid` still leaks into the normalized `Vote` at runtime** — `vote.service.ts:339`

```ts
return { ...raw, uid: uid ?? '' } as Vote;
```

The spread keeps `vote_uid` on the runtime object; the `as Vote` cast just hides it from the type system. This is the second half of Copilot [r3139222222](https://github.com/linuxfoundation/lfx-v2-ui/pull/585#discussion_r3139222222) ("consider stripping `vote_uid` from the returned object to avoid leaking an internal index field") and was present in @asithade's suggested helper shape but not in the implementation. Low risk today, but it can silently mask bugs if a future caller relies on `(vote as any).vote_uid` or if `IndexedVote` gains additional indexer-only fields. Suggested minimal fix:

```ts
const { vote_uid: _discard, ...rest } = raw;
return { ...rest, uid: uid ?? '' };
```

**2. `deleteVote` polling still types the response as `QueryServiceResponse<Vote>`** — `vote.service.ts:171`

Functionally correct (only `resources.length === 0` is checked), but inconsistent with the other three call sites that were moved to `IndexedVote`. Worth a one-liner switch so a future reader doesn't have to figure out why this one differs.

**3. No unit tests for `normalizeIndexedVote`**

The helper is pure and trivially testable. The `as Vote` cast + `|| ''` fallback is exactly the kind of edge case that benefits from a safety net. Three cases cover the behavior: (a) `uid` present, (b) `vote_uid` present, (c) both missing → warning + empty-string return.

**4. `getMyVotes` uses an ad-hoc `{ vote_uid: string }` generic** — `vote.service.ts:273-274`

That shape is `IndividualVote` (the `vote_response` indexer record already in `poll.interface.ts`). A `Pick<IndividualVote, 'vote_uid'>` would stay consistent with the "use shared types" convention the rest of this PR is embracing. Nit — not a blocker.

**5. Test plan checklist is empty**

Has the manual walkthrough (project Votes tab → Results, committee Votes tab → Results, Me lens regression, create vote, enable vote) been run against the deploy preview? The fix is clearly correct; a quick confirmation before merge closes the loop.

### Verdict

Approve once item 1 is either applied or explicitly declined as intentional. Items 2–5 are nice-to-haves and fine as follow-ups.

🤖 Generated with [Claude Code](https://claude.ai/claude-code)

**Review by asithade — DISMISSED** [review]

File: packages/shared/src/interfaces/poll.interface.ts [hunk 10]
```diff
@@ -217,6 +217,17 @@ export interface Vote {
   num_response_received?: number;
 }
 
+/**
+ * Vote resource shape as returned by the query service indexer.
+ * Uses `vote_uid` (v2 primary key) — NOT `uid`.
+ * Normalize to canonical `Vote` via VoteService.normalizeIndexedVote().
```
[the remark this answers is not in the document]
Comment (MRashad26) [reply packages/shared/src/interfaces/poll.interface.ts]: Fixed in 54b25738 — removed the reference to `VoteService.normalizeIndexedVote()` from the JSDoc. The comment now describes the normalization generically.

File: packages/shared/src/interfaces/poll.interface.ts [hunk 11]
```diff
@@ -217,6 +217,17 @@ export interface Vote {
...
+ * @see lfx-v2-voting-service/docs/indexer-contract.md
```
[the remark this answers is not in the document]
Comment (MRashad26) [reply packages/shared/src/interfaces/poll.interface.ts]: Fixed in 54b25738 — removed the `@see` reference.

File: apps/lfx-one/src/server/services/vote.service.ts [hunk 12]
```diff
@@ -317,4 +319,23 @@ export class VoteService {
...
+   * while the REST API returns `uid` — these are the same value, different field names.
+   */
+  private normalizeIndexedVote(req: Request, raw: IndexedVote): Vote {
+    const uid = raw.uid || raw.vote_uid;
+    if (!uid) {
+      logger.warning(req, 'normalize_indexed_vote', 'Indexed vote missing uid and vote_uid', {
+        name: raw.name,
+      });
+    }
+    return { ...raw, uid: uid ?? '' } as Vote;
```
[the remark this answers is not in the document]
Comment (MRashad26) [reply apps/lfx-one/src/server/services/vote.service.ts]: No code change here — the `logger.warning` on missing uid already covers the data-quality concern. Throwing would break the call sites (which expect a `Vote[]`, not an exception on bad index data), and filtering at call sites would require touching all three. The warning surfaces the issue in logs without disrupting the fetch.

[the remark this answers is not in the document]
Comment (MRashad26) [reply apps/lfx-one/src/server/services/vote.service.ts]: Fixed in 54b25738 — `vote_uid` is now stripped via destructuring before returning the canonical `Vote` shape. The ESLint `no-unused-vars` rule required an `eslint-disable-next-line` comment since the rule has no `_` prefix exemption configured.

**Comment by MRashad26** [comment]
@dealako Thanks for the thorough review. Item 1 is done — `vote_uid` is now stripped from the normalized `Vote` via destructuring in commit 54b25738. (The `no-unused-vars` rule has no `_` prefix exemption in this project's config, so the line needed an `eslint-disable-next-line` comment.)

Items 2–5 noted as nice-to-haves — happy to pick those up as a follow-up if you'd like them before merge.

**Review by dealako — APPROVED** [review]
